### PR TITLE
Feat/41 check for empty downloads

### LINKED ISSUE #41: `get_ishares_index_data()` continues without error while returning an empty table
When unable to retrieve data, the function does not raise an error, but silently returns an empty table, which causes downstream processing errors.

``` r
url <-
paste0(
  "https://www.ishares.com/uk/individual/en/products/",
  "251813/ishares-global-corporate-bond-ucits-etf/"
)
name <- "iShares Global Corporate Bond UCITS ETF <USD (Distributing)>"
as_of_date <- "20231231"
pacta.data.scraping::get_ishares_index_data(url, name, as_of_date)
#> # A tibble: 0 × 3
#> # ℹ 3 variables: base_url <chr>, index_name <chr>, as_of_date <chr>
```

<sup>Created on 2024-02-26 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

<details style="margin-bottom:10px;">
<summary>
Session info
</summary>

``` r
sessioninfo::session_info()
#> ─ Session info ───────────────────────────────────────────────────────────────
#>  setting  value
#>  version  R version 4.3.2 (2023-10-31)
#>  os       macOS Sonoma 14.2
#>  system   aarch64, darwin23.0.0
#>  ui       unknown
#>  language (EN)
#>  collate  en_US.UTF-8
#>  ctype    en_US.UTF-8
#>  tz       Europe/Belgrade
#>  date     2024-02-26
#>  pandoc   3.1.7 @ /opt/homebrew/bin/ (via rmarkdown)
#> 
#> ─ Packages ───────────────────────────────────────────────────────────────────
#>  package             * version    date (UTC) lib source
#>  cli                   3.6.2      2023-12-11 [1] CRAN (R 4.3.2)
#>  countrycode           1.5.0      2023-05-30 [1] CRAN (R 4.3.1)
#>  curl                  5.2.0      2023-12-08 [1] CRAN (R 4.3.2)
#>  digest                0.6.34     2024-01-11 [1] CRAN (R 4.3.2)
#>  dplyr                 1.1.4      2023-11-17 [1] CRAN (R 4.3.2)
#>  evaluate              0.23       2023-11-01 [1] CRAN (R 4.3.2)
#>  fansi                 1.0.6      2023-12-08 [1] CRAN (R 4.3.2)
#>  fastmap               1.1.1      2023-02-24 [1] CRAN (R 4.3.2)
#>  fs                    1.6.3      2023-07-20 [1] CRAN (R 4.3.2)
#>  generics              0.1.3      2022-07-05 [1] CRAN (R 4.3.1)
#>  glue                  1.7.0      2024-01-09 [1] CRAN (R 4.3.2)
#>  htmltools             0.5.7      2023-11-03 [1] CRAN (R 4.3.2)
#>  httr                  1.4.7      2023-08-15 [1] CRAN (R 4.3.1)
#>  jsonlite              1.8.8      2023-12-04 [1] CRAN (R 4.3.2)
#>  knitr                 1.45       2023-10-30 [1] CRAN (R 4.3.2)
#>  lifecycle             1.0.4      2023-11-07 [1] CRAN (R 4.3.1)
#>  logger                0.2.2      2021-10-19 [1] CRAN (R 4.3.1)
#>  magrittr              2.0.3      2022-03-30 [1] CRAN (R 4.3.2)
#>  pacta.data.scraping   0.1.0.9000 2024-01-18 [1] Github (RMI-PACTA/pacta.data.scraping@9d5f632)
#>  pillar                1.9.0      2023-03-22 [1] CRAN (R 4.3.1)
#>  pkgconfig             2.0.3      2019-09-22 [1] CRAN (R 4.3.1)
#>  purrr                 1.0.2      2023-08-10 [1] CRAN (R 4.3.2)
#>  R.cache               0.16.0     2022-07-21 [1] CRAN (R 4.3.1)
#>  R.methodsS3           1.8.2      2022-06-13 [1] CRAN (R 4.3.1)
#>  R.oo                  1.25.0     2022-06-12 [1] CRAN (R 4.3.1)
#>  R.utils               2.12.2     2022-11-11 [1] CRAN (R 4.3.1)
#>  R6                    2.5.1      2021-08-19 [1] CRAN (R 4.3.1)
#>  reprex                2.0.2      2022-08-17 [1] CRAN (R 4.3.1)
#>  rlang                 1.1.3      2024-01-10 [1] CRAN (R 4.3.2)
#>  rmarkdown             2.25       2023-09-18 [1] CRAN (R 4.3.1)
#>  rvest                 1.0.3      2022-08-19 [1] CRAN (R 4.3.1)
#>  selectr               0.4-2      2019-11-20 [1] CRAN (R 4.3.1)
#>  sessioninfo           1.2.2      2021-12-06 [1] CRAN (R 4.3.1)
#>  stringi               1.8.3      2023-12-11 [1] CRAN (R 4.3.2)
#>  stringr               1.5.1      2023-11-14 [1] CRAN (R 4.3.1)
#>  styler                1.10.2     2023-08-29 [1] CRAN (R 4.3.1)
#>  tibble                3.2.1      2023-03-20 [1] CRAN (R 4.3.2)
#>  tidyselect            1.2.0      2022-10-10 [1] CRAN (R 4.3.1)
#>  utf8                  1.2.4      2023-10-22 [1] CRAN (R 4.3.2)
#>  vctrs                 0.6.5      2023-12-01 [1] CRAN (R 4.3.2)
#>  withr                 3.0.0      2024-01-16 [1] CRAN (R 4.3.2)
#>  xfun                  0.41       2023-11-01 [1] CRAN (R 4.3.2)
#>  xml2                  1.3.6      2023-12-04 [1] CRAN (R 4.3.2)
#>  yaml                  2.3.8      2023-12-11 [1] CRAN (R 4.3.2)
#> 
#>  [1] /opt/homebrew/lib/R/4.3/site-library
#>  [2] /opt/homebrew/Cellar/r/4.3.2/lib/R/library
#> 
#> ──────────────────────────────────────────────────────────────────────────────
```

</details>

with `debugonce()`, I have determined that this is because the `aaData` object returned by the server is empty.
```

[ins] Browse[2]> readLines(data_path)
[1] "{\"aaData\":[]}"
Warning message:
In readLines(data_path) :
  incomplete final line found on '/var/folders/g8/1kf0nz093f3gmy0t9yhtwbq00000gn/T//Rtmp17d66K/file1300659d0db32'

```

Relates to https://github.com/RMI-PACTA/workflow.prepare.pacta.indices/issues/68

### PR DESCRIPTION
**Pull request by AlexAxthelm**
Add checks that data pulled from sources is not empty

Closes #41

### CODE REVIEW

**Review by jdhoffa — APPROVED** [review]
lgtm!

**Comment by cjyetman** [comment]
I forgot that I am the maintainer of this repo now 🤣 
I'll take a look at this tomorrow.

File: R/get_currency_exchange_rates.R [hunk 1]
```diff
@@ -32,22 +32,29 @@ get_currency_exchange_rates <-
...
 
-    httr2::request(url) %>%
+    raw_data <- httr2::request(url) %>%
       httr2::req_headers("Accept" = "application/json") %>%
       httr2::req_timeout(max_seconds) %>%
       httr2::req_retry(
         max_seconds = max_seconds,
         is_transient = ~ httr2::resp_content_type(.x) != "application/json"
-      ) %>%
+        ) %>%
```
Comment (cjyetman) [inline R/get_currency_exchange_rates.R:41]:
```suggestion
      ) %>%
```

File: R/get_currency_exchange_rates.R [hunk 2]
```diff
@@ -42,12 +42,19 @@ get_currency_exchange_rates <-
       httr2::req_error(
         is_error = ~ httr2::resp_content_type(.x) != "application/json",
         body = ~ cli::format_inline("IMF API did not return a valid response after {.val {max_seconds}} seconds.")
-      ) %>%
+        ) %>%
```
Comment (cjyetman) [inline R/get_currency_exchange_rates.R:45]:
```suggestion
      ) %>%
```

File: R/get_index_regions.R [hunk 3]
```diff
@@ -34,6 +34,10 @@ get_index_regions <- function() {
     ct$get("JSON.stringify(chartDataDMJson)"),
     flatten = TRUE
   )
+  if (length(dm_data$categories) == 0) {
```
Comment (cjyetman) [inline R/get_index_regions.R:37]:
```suggestion
  if (length(dm_data$categories) == 0L) {
```

File: R/get_index_regions.R [hunk 4]
```diff
@@ -72,6 +76,10 @@ get_index_regions <- function() {
     ct$get("JSON.stringify(chartDataEMJson)"),
     flatten = TRUE
   )
+  if (length(data$categories) == 0) {
```
Comment (cjyetman) [inline R/get_index_regions.R:79]:
```suggestion
  if (length(data$categories) == 0L) {
```

File: R/get_ishares_index_data.R [hunk 5]
```diff
@@ -44,6 +44,11 @@ get_ishares_index_data <- function(url, name, as_of_date) {
     # JSON string contains (illegal) UTF8 byte-order-mark!
     jsonlite::fromJSON(data_path)$aaData
   )
+  if(length(raw_data) == 0) {
```
Comment (cjyetman) [inline R/get_ishares_index_data.R:47]:
```suggestion
  if(length(raw_data) == 0L) {
```

**Review by cjyetman — APPROVED** [review]

**Pull request merged**